File: Занятие 1 - Сложность, тестирование, особые случаи/D_Слоны_и_ладьи.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    vector<vector<int>> board(8, vector<int>(8, 0)); // Массив заполенный нулями: {0, 0, 0.......}

    for (int i = 0; i < 8; ++i) {
        string line;
        cin >> line; // получаю строчные данные, разбиваю их на символы и записываю цифры,, в соответсвии с символом
        for (int j = 0; j < 8; ++j) {
            if (line[j] == '*') board[i][j] = 0;
            if (line[j] == 'B') board[i][j] = 1;
            if (line[j] == 'R') board[i][j] = 2;
        }
    } // на выходе массив, заполненный 0, 1 и 2 в соответсвии с символами

    for (int i = 0; i < 8; ++i) {
        for (int j = 0; j < 8; ++j) {
            if (board[i][j] == 2) {
                for (int k = i + 1; k < 8; ++k) {
                    if (board[k][j] == 0) board[k][j] = -1; // если ячейка не занята и не обработана, т.е. == 0, заменяем ее на -1, обработанную
                    else if (board[k][j] == 1 || board[k][j] == 2) break; // если ячейка занята фигурой, останавливаем цикл
                }
                for (int k = i - 1; k >= 0; --k) {
                    if (board[k][j] == 0) board[k][j] = -1;
                    else if (board[k][j] == 1 or board[k][j] == 2) break; // C++ понимает и английское обозначение операторов (можно писать и "||" и "or")
                }
                for (int k = j + 1; k < 8; ++k) {
                    if (board[i][k] == 0) board[i][k] = -1;
                    else if (board[i][k] == 1 || board[i][k] == 2) break;
                }
                for (int k = j - 1; k >= 0; --k) {
                    if (board[i][k] == 0) board[i][k] = -1;
                    else if (board[i][k] == 1 || board[i][k] == 2) break;
                }
            }
            if (board[i][j] == 1) {
                for (int k = i + 1, p = j + 1; k < 8 && p < 8; ++k, ++p) {
                    if (board[k][p] == 0) board[k][p] = -1;
                    else if (board[k][p] == 1 || board[k][p] == 2) break;
                }
                for (int k = i + 1, p = j - 1; k < 8 && p >= 0; ++k, --p) {
                    if (board[k][p] == 0) board[k][p] = -1;
                    else if (board[k][p] == 1 || board[k][p] == 2) break;
                }
                for (int k = i - 1, p = j + 1; k >= 0 && p < 8; --k, ++p) {
                    if (board[k][p] == 0) board[k][p] = -1;
                    else if (board[k][p] == 1 || board[k][p] == 2) break;
                }
                for (int k = i - 1, p = j - 1; k >= 0 && p >= 0; --k, --p) {
                    if (board[k][p] == 0) board[k][p] = -1;
                    else if (board[k][p] == 1 || board[k][p] == 2) break;
                }
            }
        }
    }

    int counter = 0; // считаю оставшиеся нули в массиве
    for (int i = 0; i < 8; ++i) {
        for (int j = 0; j < 8; ++j) {
            if (board[i][j] == 0) counter++;
        }
    }
    cout << counter << endl;
}

File: Занятие 1 - Сложность, тестирование, особые случаи/D. Слоны и ладьи.py
board = [str(input()) for _ in range(8)]

deleted = []

counter=64

for i in range(8):
    for j in range(8):
        if board[i][j] == 'R':
            counter -= 1
            if j != 0:
                for z in range(j-1, -1, -1):
                    if board[i][z] == '*':
                        if (i, z) not in deleted:
                            counter -= 1
                            deleted.append((i, z))
                    else:
                        break
            if j != 7:
                for x in range(j+1, 8, 1):
                    if board[i][x] == '*':
                        if (i, x) not in deleted:
                            counter -= 1
                            deleted.append((i, x))
                    else:
                        break
            if i != 0:
                for c in range(i-1, -1, -1):
                    if board[c][j] == '*':
                        if (c, j) not in deleted:
                            counter -= 1
                            deleted.append((c, j))
                    else:
                        break
            if i!=7:
                for v in range(i+1, 8, 1):
                    if board[v][j] == '*':
                        if (v, j) not in deleted:
                            counter -= 1
                            deleted.append((v, j))
                    else:
                        break
                
        elif board[i][j] == 'B':
            counter -= 1
            if i != 0:
                if j != 0:
                    i_2 = i
                    j_2 = j
                    while i_2 > 0 and j_2 > 0:
                        i_2 -= 1
                        j_2 -= 1
                        if board[i_2][j_2] == '*':
                            if (i_2, j_2) not in deleted:
                                counter -= 1
                                deleted.append((i_2, j_2))
                        else:
                            break
                if j != 7:
                    i_2 = i
                    j_2 = j
                    while i_2 > 0 and j_2 < 7:
                        i_2 -= 1
                        j_2 += 1
                        if board[i_2][j_2] == '*':
                            if (i_2, j_2) not in deleted:
                                counter -= 1
                                deleted.append((i_2, j_2))
                        else:
                            break
            if i != 7:
                if j != 0:
                    i_2 = i
                    j_2 = j
                    while i_2 < 7 and j_2 > 0:
                        i_2 += 1
                        j_2 -= 1
                        if board[i_2][j_2] == '*':
                            if (i_2, j_2) not in deleted:
                                counter -= 1
                                deleted.append((i_2, j_2))
                        else:
                            break
                if j != 7:
                    i_2 = i
                    j_2 = j
                    while i_2 < 7 and j_2 < 7:
                        i_2 += 1
                        j_2 += 1
                        if board[i_2][j_2] == '*':
                            if (i_2, j_2) not in deleted:
                                counter -= 1
                                deleted.append((i_2, j_2))
                        else:
                            break
        else:
            pass

print(counter)
